support for parsing arbitrary weight numbers and stretch percentages

russaa opened the pull request:
adds support for parsing extended capabilities defined in _CSS Fonts Module Level 4_ for

 * `font-weight`: allow arbitrary numbers from 1 to 1000
    * https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight
    * https://drafts.csswg.org/css-fonts-4/#valdef-font-weight-number
 * `font-stretch`: allow (positive) percentage numbers
  * https://developer.mozilla.org/en-US/docs/Web/CSS/font-stretch
  * https://drafts.csswg.org/css-fonts-4/#valdef-font-stretch-percentage

this PR also extends parsing of valid numbers:
namely considering leading `+` and `-` when parsing numbers
See also
https://developer.mozilla.org/en-US/docs/Web/CSS/integer
https://drafts.csswg.org/css-values-3/#numeric-types

jedmao commented on src/index.ts:
```diff
@@ -94,7 +110,18 @@ export default function parseCSSFont(value: string) {
+		} else if (prelimStretchNum) {
+			font.size = font.stretch;
+			font.stretch = 'normal';
+			font.family = cssListHelpers.splitByCommas(tokens.join(' ')).map(unquote) as string[];
```
Ditto on the `as string[]`.

jedmao commented on src/index.ts:
```diff
@@ -119,6 +146,23 @@ function parseLineHeight(value: string) {
 	if (parsed.toString() === value) {
 		return parsed;
+	} else {
+		const match = /^(\+|-)?(\.)?/.exec(value) as RegExpMatchArray;
```
We can probably reuse the above-mentioned regex that I asked you to cache here.

Also, the TypeScript compiler will infer the type, so you shouldn't need `as RegExpMatchArray` here.

russaa replied:
without the cast, the compiler will complain that the `match` could be `null`:
but the regular expression will always match, since it has 2 optional parts.

So either have this cast, or add a not-`null` check to convince the compiler that using `match[1]` is OK -- I decided to use the cast, but if you prefer the not-`null` check, I will change it.

Similar for those other casts above:
it's i.m.o. a matter of taste, if you use the cast or add an explicit check -- I prefer the cast for instances where I know that the explicit check is unnecessary, as it does not add any JavaScript code after the compilation

jedmao replied:
Why are you so sure the value wouldn't be `null`?

russaa replied:
it checks for (a) a beginning "^" and (b) two optional groups, so as long as the string has a beginning it will match

jedmao replied:
I'm not sure what's wrong w/ your compiler or at what point it's giving you these errors, but I can tell you that the TypeScript playground has no problems with the following code:
```ts
function foo(value: string) {
    const match = /^(foo)?(bar)?/.exec(value);
    return match[1];
}
```
If I mouse-over `const match` it tells me that `match` is a `RegExpExecArray`.
![image](https://user-images.githubusercontent.com/1058243/44050842-cf185a3c-9efd-11e8-82be-21bc7de296b3.png)
Are you seeing something different? Note that `RegExpExecArray` and `RegExpMatchArray` are two different interfaces.

russaa replied:
no you are right, it should be `RegExpExecArray` (since this was only for silencing the compiler I did not look closely enough)

still: for `exec()` my compiler shows me a return value of `RegExpExecArray | null`, so without the cast there would be an explicit check for non-`null`

jedmao commented on src/index.ts:
```diff
@@ -119,6 +146,23 @@ function parseLineHeight(value: string) {
+		const match = /^(\+|-)?(\.)?/.exec(value) as RegExpMatchArray;
+		let val: string = value;
+		if (match[1] === '+') {
+			val = val.substring(1);
```
This shouldn't be needed once we deconstruct the match into `[sign, value, percent]`.

jedmao commented on src/index.ts:
```diff
@@ -119,6 +146,23 @@ function parseLineHeight(value: string) {
+			val = val.substring(1);
+		}
+		if (match[2] === '.') {
+				// NOTE although not specifically prohibited, we do not consider negative numbers for line-height valid:
```
Were all these comments left here on purpose? Were these issues ever resolved?

russaa replied:
yes, I left this comment on purpose, although I will happily remove it:

The underlying reason for the comment was that I am not really sure why there is a conversion from `string` to  `number` at this point.

I did not want to touch the `line-height` parsing code at all, but the test framework requires the conversion of the `string` values to `number` (otherwise the extended number tests could not have been added, or would have needed to exclude tests where `line-height` was part of the test input).

The comment is wrong in that really [negative numbers are not allowed for `line-height`](https://drafts.csswg.org/css2/visudet.html#propdef-line-height), but as I said, I did not want to mess around too much with this code as I was not really sure what the intention was here:
why convert to numbers? and why omit the units?

jedmao commented on src/index.ts:
```diff
@@ -119,6 +146,23 @@ function parseLineHeight(value: string) {
 	const parsed = parseFloat(value);
 	if (parsed.toString() === value) {
 		return parsed;
+	} else {
```
Another superfluous `else`.

jedmao replied:
This `else` can still be removed.

russaa commented on src/helpers.ts:
```diff
@@ -1,8 +1,39 @@
+
+function isOnlyNumber(value: string, isPercent: boolean): boolean {
+	const match = /^(\+|-)?(\d|\.)/.exec(value);
+	if (match && (!isPercent || /%$/.test(value))) {
```
the meaning of `isPercent` is something like "only return `true` if it is number and also has a percent sign", so the naming is probably a bit of a poor choice, maybe `onlyPercent` would make it more clear:

it means, either accept any numbers or, if `isPercent` is `true`, require that a percent sign follows the number

jedmao replied:
The "and also" part is the confusing part and violates the SRP. Try to make your functions do one and only one thing to clear up any confusion.

russaa commented on src/index.ts:
```diff
@@ -58,12 +59,21 @@ export default function parseCSSFont(value: string) {
+				font.weight = token;
+				continue;
+			} else {
+				const num = parseFloat(token);
```
I don't think, I understand this comment:
I do need a number when I want to test `< 1` and `> 1000`, or how would you do it?

jedmao replied:
Yeah that makes sense. You're good here.

russaa commented on src/index.ts:
```diff
@@ -94,7 +110,18 @@ export default function parseCSSFont(value: string) {
 				throw error('Missing required font-family.');
 			}
-			font.family = cssListHelpers.splitByCommas(tokens.join(' ')).map(unquote);
+			font.family = cssListHelpers.splitByCommas(tokens.join(' ')).map(unquote) as string[];
```
without this cast, I get an error `Type '{}[]' is not assignable to type 'string[] | undefined'. Type '{}[]' is not assignable to type 'string[]'. Type '{}' is not assignable to type 'string'.`

jedmao replied:
Let me take a look at this locally.

jedmao replied:
I can safely remove `as string[]` locally w/o issue. No lint issues and compiles just fine. I'm curious, what editor are you using and what version of TypeScript do you see if you run `tsc --version`? You should definitely try upgrading just to be sure.

russaa replied:
thanks for the tip: for some reason `npm install` did put `typescript` version `2.3.4` into `/node_modules`

russaa commented on src/index.ts:
```diff
@@ -94,7 +110,18 @@ export default function parseCSSFont(value: string) {
-			font.family = cssListHelpers.splitByCommas(tokens.join(' ')).map(unquote);
+			font.family = cssListHelpers.splitByCommas(tokens.join(' ')).map(unquote) as string[];
+			if (prelimStretchNum) {
+				const num = parseFloat(font.stretch as string);
```
similar to the previous comment: yes, it always be a string, but without the cast, the compiler complains `Argument of type 'string | undefined' is not assignable to parameter of type 'string'. Type 'undefined' is not assignable to type 'string'.`

russaa commented:
I will work on those changes & post an update

for some that were not fully clear to me, I added comments

russaa commented on src/helpers.ts:
```diff
@@ -1,8 +1,39 @@
+}
+
+function isOnlyNumber(value: string, isPercent: boolean): boolean {
+	const match = /^(\+|-)?(\d|\.)/.exec(value);
```
I am not sure about this (changing the regular expression):

 * as to the first question, this regular expression is only a pre-test: any non-number value would get rejected at `parseFloat(val).toString() === val`

 * this regular expression is also used for determining if there is leading zero missing for fraction-only values (e.g. `.24`): in this case a zero has to be prepended, so that the test `parseFloat(val).toString() === val` works

-> so changing the regular expression would also require to re-implement the test (for checking if the string really only contains a number, or something more, or something else entirely)

jedmao replied:
Either way, for readability, changing the variable names to `sign`, `value` and `percent` would be much more clear.

jedmao commented on src/helpers.ts:
```diff
@@ -1,8 +1,39 @@
+		|| isOnlyNumber(value, true);
+}
+
+function isOnlyNumber(value: string, isPercent: boolean): boolean {
```
Also, instead of `isPercent` here, would it make more sense to name it `isPercentAccepted`? Or am I misunderstanding something?

russaa commented on src/helpers.ts:
```diff
@@ -1,8 +1,39 @@
+				val = '0' + val;
+			}
+		}
+		return parseFloat(val).toString() === val;
```
the purpose here is to force evaluating the complete string as a number:
`parseInt` will return a number, even if there is something else in the string, as long as there is a number at the beginning; `parseFloat` will also parse integer values correctly, just won't allow something else (non-number) in the string

jedmao commented on src/index.ts:
```diff
@@ -58,12 +59,21 @@ export default function parseCSSFont(value: string) {
+			if (fontWeightKeywords.indexOf(token) !== -1) {
+				font.weight = token;
+				continue;
+			} else {
```
You're right. Ignore request #3.

russaa commented:
I think I addressed most of your concerns except for some few were I decided on a slightly different solution:

mainly
 * in `helpers.ts` not extracting/using regular expression `const cssNumberPattern = /^(\+|-)?([\d\.]+)(%)?$/;` etc.
 * in `index.ts` I did change `parseLineHeight()` only slightly: I want to hear your general feedback first

jedmao commented on src/index.ts:
```diff
@@ -119,6 +140,19 @@ function parseLineHeight(value: string) {
 	if (parsed.toString() === value) {
 		return parsed;
+	} else {
+		const match = /^(\+)?(\.)?/.exec(value) as RegExpExecArray;
```
Let's add the following to the `tsconfig.json` compiler options:
```json
		"downlevelIteration": true,
```
And then you can remove this `as RegExpExecArray;` here.

jedmao commented on src/index.ts:
```diff
@@ -120,5 +141,17 @@ function parseLineHeight(value: string) {
 	if (parsed.toString() === value) {
 		return parsed;
 	}
+	const match = /^(\+)?(\.)?/.exec(value);
```
OK the more I look at this logic, the more I think installing [`parse-css-dimension`](https://github.com/jedmao/parse-css-dimension) might be useful in replacing this logic. What do you think?

jedmao replied:
No broken tests with the following implementation:
```ts
function parseLineHeight(value: string) {
	try {
		const parsed = parseCssDimension(value);
		if (parsed.type === 'number') {
			return parsed.value;
		}
	} catch (err) {
		// noop
	}
	return value;
}
```

russaa replied:
yes that would probably be better -- ideally the parsing should also be extended to reject negative values for line height (and invalid dimensions/units)

jedmao replied:
I see what you mean. Looks like there's a validation aspect that needs to be employed here. I just tried this:
```css
.foo {
  line-height: -1;
}
```
on http://jigsaw.w3.org/css-validator/#validate_by_input and got the following error:
> Value Error : line-height -1 negative values are not allowed
We could easily add the following:
```js
if (parsed.value < 0) {
    throw error(`Value Error: line-height: ${parsed.value}; negative values are not allowed.`);
}
```

jedmao commented on src/helpers.ts:
```diff
@@ -1,8 +1,50 @@
+	return false;
+}
+
+function isOnlyNumber(value: string): boolean {
```
[`parse-css-dimension`](https://github.com/jedmao/parse-css-dimension) might also be useful here.

russaa replied:
I am not sure:
the "allow number only" parsing is not really provided by `parse-css-dimension`, or more precisely, it does not make sure that the input-string does not contain more than the number and/or the the unit

E.g. for `"43asdfa7rem"` it will return a valid result with number `43` and unit `"rem"`

jedmao replied:
Aha. Sounds like it needs a strict mode then. That can be arranged.

russaa replied:
yes, then most of this PR should be removed (or moved to `parse-css-dimension` where appropriate) & replaced by using `parse-css-dimension`

Would you start on `parse-css-dimension`, e.g. how it should be invoked for "strict mode"?
e.g. maybe something like `parseCssDimension('42dpi', 'strict')` or maybe `parseCssDimension('42dpi', {strict: true})`, or ...?

I would then transfer/add parts from this PR for the strict-number testing as much as possible

jedmao replied:
I think we should go with an options hash of `{ strict: true }`.

jedmao replied:
@russaa see https://github.com/jedmao/parse-css-dimension/issues/4. Also, [I'm updating that project to TypeScript/AVA](https://github.com/jedmao/parse-css-dimension/pull/3) like this one. Feel free to submit a PR for https://github.com/jedmao/parse-css-dimension/issues/4.

jedmao commented on src/index.test.ts:
```diff
@@ -6,6 +6,57 @@ const fontWeightKeywords = require('css-font-weight-keywords');
 const fontStyleKeywords = require('css-font-style-keywords');
 const fontStretchKeywords = require('css-font-stretch-keywords');
 
+const numberValues = [
```
I think you'll find that many of these tests cases are covered [here](https://github.com/jedmao/parse-css-dimension/blob/master/test/index.js), so as not to reinvent the wheel.

jedmao commented:
Closing for now, due to inactivity.

jedmao commented:
@russaa I have a 🎁 for you: https://www.npmjs.com/package/parse-css-number

russaa commented:
@jedmao very nice  :smiley: using `JSON.parse` :thumbsup:
